server/src/main/java/org/opensearch/common/xcontent/JsonToStringXContentParser.java

```diff
@@ -86,8 +83,22 @@ private void parseToken(StringBuilder path, String currentFieldName) throws IOEx
             StringBuilder parsedFields = new StringBuilder();
 
             if (this.parser.currentToken() == Token.FIELD_NAME) {
-                path.append(DOT_SYMBOL + currentFieldName);
-                this.keyList.add(currentFieldName);
+                path.append(DOT_SYMBOL).append(currentFieldName);
+                int dotIndex = currentFieldName.indexOf(DOT_SYMBOL);
+                String fieldNameSuffix = currentFieldName;
+                // The field name may be of the form foo.bar.baz
+                // If that's the case, each "part" is a key.
+                while (dotIndex >= 0) {
```

Comment:
Should we put a limit on how big the keyList is?  Or is that guaranteed to be reasonable because we would have rejected the request if the bulk size was too large.

Reply:
+1 to @kkmr 's highlight, we should consult `MappingService::INDEX_MAPPING_DEPTH_LIMIT_SETTING` since we do depth analysis here  (and probably `MappingService::INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING`)

Reply:
IIRC, when @mingshl first tested flat object, she found that it supports very long, very deep sub fields.

From https://opensearch.org/docs/latest/field-types/supported-field-types/flat-object/:
> The maximum field value length in the dot notation is 2^24 − 1.

Because it all gets stored in the underlying terms, we don't need to worry about a mapping explosion, the way we do for object or nested fields.

Reply:
Ah fair point, thanks @msfroh

Reply:
Thanks @msfroh